quickwit/quickwit-ingest/src/ingest_api_service.rs

```diff
@@ -241,6 +241,8 @@ impl IngestApiService {
             .await?;
 
         let memory_usage = self.queues.memory_usage();
+        // TODO trinity: i got `an attempt to subtract with overflow` once here. should investigate
```

Comment:
This is ingest v1. Did you run your tests against ingest v1 or v2? It would make sense that you observed a memory usage spike on ingest v1 when many requests are in flight because there is no concurrency limit on ingest v1.

On the other hand, in ingest v2, the router does limit the number of requests in flight.

Reply:
while testing, i did a good mix of v1 and v2, iirc this error was with v1

Reply:
We don't care about ingest v1 at all. If this fix is still working on v2, do we have an understanding as to why the backpressure from the router is not applied to the REST API?

Reply:
that comment is unrelated to the rest of the pr.
my understanding of what happens with the backpressure not working is the following:
- many requests arrives concurrently (let's say 1k). We allocate a buffer for each body, and start receiving the requests.
- once the body of a request is fully read, the route gets called
- the route calls the inner ingest api code, which for the vast majority of requests, will reply rate_limited
at this point, we have already allocated (and at least partially filled) all those buffers. This PR adds a semaphore sooner, just before allocating the buffer for a request body, so that we don't have too much such body around at any given instant. The requests it rejects would be rejected later, when entering ingest api code, but that way we don't pay for the memory usage of these requests

Reply:
Ok, in other words, we would allocate HTTP bodies faster than we reject them in the router.